File: telegram-py-webhook/telegram.ts
import type {
    BotCommand,
    File,
    InlineKeyboardMarkup,
    ParseMode,
} from "npm:@types/node-telegram-bot-api";

const api = "https://api.telegram.org";

import { nice } from "./pprint.ts";

const nicer = (v: unknown) => (v ? nice(v) : "");

export class Telegram {
    token: string;

    constructor(token: string) {
        this.token = token;
    }

    file(path: string) {
        return `${api}/file/bot${this.token}/${path}`;
    }

    async answerCallbackQuery(callback_query_id: string, text?: string) {
        return await this._command("answerCallbackQuery", {
            callback_query_id,
            text,
        });
    }

    async sendMessage(
        chat_id: number,
        text: string,
        reply_markup?: InlineKeyboardMarkup,
        parse_mode?: ParseMode
    ) {
        return await this._command("sendMessage", {
            chat_id,
            text,
            reply_markup,
            parse_mode,
        });
    }

    async editMessageText(chat_id: number, message_id: number, text: string) {
        return await this._command("editMessageText", {
            chat_id,
            message_id,
            text,
        });
    }

    async sendPhoto(
        chat_id: number,
        photo: string,
        reply_markup?: InlineKeyboardMarkup
    ) {
        return await this._command("sendPhoto", {
            chat_id,
            photo,
            reply_markup,
        });
    }

    async sendAudio(
        chat_id: number,
        audio: string,
        reply_markup?: InlineKeyboardMarkup
    ) {
        return await this._command("sendAudio", {
            chat_id,
            audio,
            reply_markup,
        });
    }

    async getFile(file_id: string): Promise<File> {
        return await this._command("getFile", { file_id });
    }

    async setWebhook(url: string) {
        return await this._command("setWebhook", {
            url,
            drop_pending_updates: true,
        });
    }

    async deleteWebhook() {
        return await this._command("deleteWebhook");
    }

    async getWebhookInfo() {
        return await this._command("getWebhookInfo");
    }

    async setMyCommands(commands: BotCommand[]) {
        return await this._command("setMyCommands", { commands });
    }

    async _command(cmd: string, data?: Record<string, unknown>) {
        const url = `${api}/bot${this.token}/${cmd}`;
        // deno-lint-ignore no-constant-condition
        if (false) console.log("TELEGRAM REQUEST", url, nicer(data));
        try {
            let response: Response;
            if (data) {
                response = await fetch(url, {
                    method: "POST",
                    headers: { "Content-Type": "application/json" },
                    body: JSON.stringify(data),
                });
            } else response = await fetch(url);
            if (!response.ok) {
                const msg = `error: ${response.status} ${response.statusText}`;
                throw new Error(msg);
            }
            const result = await response.json();
            if (!result.ok) {
                console.error("TELEGRAM ERROR", nicer(result));
                const msg = `${result.error_code} ${result.description || ""}`;
                throw new Error(msg);
            }
            return result.result;
        } catch (e: unknown) {
            console.error("TELEGRAM REQUEST", url, nicer(data));
            if (e instanceof Error) console.error("ERROR", e.message);
            else console.error("ERROR", e);
        }
    }
}

File: telegram-py-webhook/telegram_.py
import os
from typing import Any, cast

import requests

API = "https://api.telegram.org"


class Telegram:
    def __init__(self, token: str) -> None:
        self.token = token

    def file(self, path: str) -> str:
        return f"{API}/file/bot{self.token}/{path}"

    def answer_callback_query(
        self,
        callback_query_id: str,
        text: str | None = None,
    ) -> None:
        self._command(
            "answerCallbackQuery",
            {"callback_query_id": callback_query_id, "text": text},
        )

    def send_message(
        self,
        chat_id: int,
        text: str,
        reply_markup: dict[str, Any] | None = None,
        parse_mode: str | None = None,
    ) -> dict[str, Any]:
        data = {"chat_id": chat_id, "text": text}
        if reply_markup:
            data["reply_markup"] = reply_markup
        if parse_mode:
            data["parse_mode"] = parse_mode

        response = self._command("sendMessage", data)
        print(response["message_id"])
        return response

    def edit_message_text(
        self,
        chat_id: int,
        message_id: int,
        text: str,
    ) -> dict[str, Any]:
        return self._command(
            "editMessageText",
            {"chat_id": chat_id, "message_id": message_id, "text": text},
        )

    def send_photo(
        self,
        chat_id: int,
        photo: str,
        reply_markup: dict | None = None,
    ) -> None:
        data = {"chat_id": chat_id, "photo": photo}
        if reply_markup:
            data["reply_markup"] = reply_markup
        self._command("sendPhoto", data)

    def send_audio(
        self,
        chat_id: int,
        audio: str,
        reply_markup: dict | None = None,
    ) -> None:
        data = {"chat_id": chat_id, "audio": audio}
        if reply_markup:
            data["reply_markup"] = reply_markup
        self._command("sendAudio", data)

    def send_video(
        self,
        chat_id: int,
        video: str,
        reply_markup: dict | None = None,
    ) -> None:
        data = {"chat_id": chat_id, "video": video}
        if reply_markup:
            data["reply_markup"] = reply_markup
        self._command("sendVideo", data)

    def getFile(self, file_id: str) -> dict:
        return self._command("getFile", {"file_id": file_id})

    def set_webhook(self, url: str, secret_token: str | None = None) -> None:
        secret_token = os.environ.get("TELEGRAM_SECURE_TOKEN", secret_token)
        self._command(
            "setWebhook",
            {
                "url": url,
                "secret_token": secret_token,
                "drop_pending_updates": True,
            },
        )

    def delete_webhook(self) -> None:
        self._command("deleteWebhook")

    def get_webhook_info(self) -> dict[str, str]:
        return self._command("getWebhookInfo")

    def get_me(self) -> dict[str, str]:
        return self._command("getMe")

    def set_my_commands(self, commands: list[dict[str, str]]) -> None:
        self._command("setMyCommands", {"commands": commands})

    def _command(self, cmd: str, data: dict | None = None) -> dict[str, Any]:
        url = f"{API}/bot{self.token}/{cmd}"
        if data:
            response = requests.post(
                url,
                headers={"Content-Type": "application/json"},
                json=data,
            )
        else:
            response = requests.get(url)
        response.raise_for_status()
        result: dict[str, str | dict[str, Any]] = response.json()
        ok = result.get("ok", False)
        if not ok:
            raise Exception("TELEGRAM ERROR", result)
        return cast(dict[str, Any], result["result"])
